File: 2018 ICPC Vietnam National Programming Contest/communicatingstrategy.pas
program Main;

const
  MOD_VAL = 1000000007;
  INF = 1 shl 60;

type
  LongIntArray = array of int64;
  LongInt2DArray = array of array of int64;

function modinverse(a, b: int64): int64;
begin
  if 1 < a then
    modinverse := b - modinverse(b mod a, a) * b div a
  else
    modinverse := 1;
end;

procedure clean(var n: int64);
begin
  while n < MOD_VAL do
    n := n + MOD_VAL;
  n := n mod MOD_VAL;
end;

var
  n, i, j: int64;
  t1, t2, val: int64;
  memo: LongInt2DArray;
  ans: array of int64;
  ans_size: integer;

begin
  readln(n);
  
  SetLength(memo, n + 1);
  for i := 0 to n do
  begin
    SetLength(memo[i], n + 1);
    for j := 0 to n do
      memo[i][j] := 0;
  end;
  
  writeln('1 1 1');
  flush(output);
  readln(t1, t2);
  memo[1][1] := t2;
  
  SetLength(ans, n);
  ans_size := 0;
  ans[ans_size] := t2;
  ans_size := ans_size + 1;
  
  for i := 2 to n do
  begin
    writeln('1 1 ', i);
    flush(output);
    readln(t1, t2);
    
    if t1 = 1 then
    begin
      val := t2 - memo[i - 1][1];
      clean(val);
    end
    else if t1 = i then
    begin
      val := t2 * modinverse(memo[i - 1][i - 1], MOD_VAL);
      clean(val);
    end
    else
    begin
      val := (t2 - memo[i - 1][t1]);
      clean(val);
      val := val * modinverse(memo[i - 1][t1 - 1], MOD_VAL);
      clean(val);
    end;
    
    for j := 1 to i do
    begin
      if j = 1 then
      begin
        memo[i][j] := memo[i - 1][j] + val;
        clean(memo[i][j]);
      end
      else if j = i then
      begin
        memo[i][j] := memo[i - 1][j - 1] * val;
        clean(memo[i][j]);
      end
      else
      begin
        memo[i][j] := memo[i - 1][j] + memo[i - 1][j - 1] * val;
        clean(memo[i][j]);
      end;
    end;
    
    if ans_size < n then
    begin
      ans[ans_size] := val;
      ans_size := ans_size + 1;
    end;
  end;
  
  write('2 ');
  for i := 0 to ans_size - 1 do
    write(ans[i], ' ');
  writeln;
  flush(output);
end.

File: 2018 ICPC Vietnam National Programming Contest/communicatingstrategy.c
#include <stdio.h>
#include <stdlib.h>

#define MOD 1000000007LL
#define INF ((long long)1 << 60)

long long modinverse(long long a, long long b) {
    return 1 < a ? b - modinverse(b % a, a) * b / a : 1;
}

void clean(long long* n) {
    while (*n < MOD) *n += MOD;
    *n %= MOD;
}

int main() {
    long long n;
    scanf("%lld", &n);

    long long** memo = (long long**)malloc((n + 1) * sizeof(long long*));
    for (int i = 0; i <= n; i++) {
        memo[i] = (long long*)malloc((n + 1) * sizeof(long long));
        for (int j = 0; j <= n; j++) {
            memo[i][j] = 0;
        }
    }

    printf("1 1 1\n");
    fflush(stdout);
    long long t1, t2;
    scanf("%lld %lld", &t1, &t2);
    memo[1][1] = t2;

    long long* ans = (long long*)malloc(n * sizeof(long long));
    int ans_size = 0;
    ans[ans_size++] = t2;

    for (long long i = 2; i <= n; i++) {
        printf("1 1 %lld\n", i);
        fflush(stdout);
        scanf("%lld %lld", &t1, &t2);

        long long val;

        if (t1 == 1) {
            val = t2 - memo[i - 1][1];
            clean(&val);
        } else if (t1 == i) {
            val = t2 * modinverse(memo[i - 1][i - 1], MOD);
            clean(&val);
        } else {
            val = (t2 - memo[i - 1][t1]);
            clean(&val);
            val *= modinverse(memo[i - 1][t1 - 1], MOD);
            clean(&val);
        }

        for (int j = 1; j <= i; j++) {
            if (j == 1) {
                memo[i][j] = memo[i - 1][j] + val;
                clean(&memo[i][j]);
            } else if (j == i) {
                memo[i][j] = memo[i - 1][j - 1] * val;
                clean(&memo[i][j]);
            } else {
                memo[i][j] = memo[i - 1][j] + memo[i - 1][j - 1] * val;
                clean(&memo[i][j]);
            }
        }

        ans[ans_size++] = val;
    }

    printf("2 ");
    for (int i = 0; i < ans_size; i++) {
        printf("%lld ", ans[i]);
    }
    printf("\n");
    fflush(stdout);

    for (int i = 0; i <= n; i++) {
        free(memo[i]);
    }
    free(memo);
    free(ans);

    return 0;
}